### PR TITLE
Orca:  Customized collate_fn in dataloader

### PR DESCRIPTION
**Pull request by leonardozcm**
### 1. Why the change?

For third-party dataloader implementations, we can use a `collate_fn` wrapper to implement custom data processing like elements alignment:
```
def reduceWrapper(func):
    def reduceIterElement(batch):
        batch=func(batch)
        assert len(batch)==5, "Should yield inputs, labels, index, time, meta in dataloader"
        return batch[0], batch[1]
    return reduceIterElement

def train_loader_creator(config, batch_size):
    train_loader = loader.construct_loader(config, "train")
    loader.shuffle_dataset(train_loader, 0)
    train_loader.collate_fn = reduceWrapper(train_loader.collate_fn)
    return train_loader
```
It should be more natural than just integrade their code into users' implementation.

~~And it can also collate elements when `data` is dataframe or Xshards for generic usages.~~

https://github.com/intel-analytics/BigDL/issues/4410

### 2. User API changes

We add a new function type argument called `recollate_fn` to estimator.fit(). 

`recollate_fn` takes a batch of dataloader as input and ouput processed batch:
```
def customized_collate_fn(batch):
       *features, labels = batch
       new_features = process_features(features)
       new_labels = process_labels(labels)
       return new_features, new_labels
```

When a dataloader is passed to fit(), it's original `collate_fn` will be wrapped around with `recollate_fn`

By default `recollate_fn` does nothing.

### 3. Summary of the change 

Enable `recollate_fn` in bigdl, ray and spark backend.  

Before this change if we need to collate dataloader manually:
```
def reduceWrapper(func):
    def reduceIterElement(batch):
        batch=func(batch)
        print("reduce called")
        assert len(batch)==5, "Should yield inputs, labels, index, time, meta in dataloader"
        return batch[0], batch[1]
    return reduceIterElement

def train_loader_creator(config, batch_size):
    train_loader = loader.construct_loader(config, "train")
    loader.shuffle_dataset(train_loader, 0)
    train_loader.collate_fn = reduceWrapper(train_loader.collate_fn)
    return train_loader

def validation_data_creator(config, batch_size):
      val_loader = loader.construct_loader(config,"val")
      val_loader.collate_fn = reduceWrapper(val_loader.collate_fn)
      return val_loader

orca_estimator.fit(data=train_loader_creator,
                               validation_data=validation_data_creator,
                               batch_size=cfg.TRAIN.BATCH_SIZE,
                               epochs=cfg.SOLVER.MAX_EPOCH)
```

now:
```
def reduceIterElement(batch):
    print("reduce called")
    assert len(batch)==5, "Should yield inputs, labels, index, time, meta in dataloader"
    return batch[0], batch[1]

orca_estimator.fit(data=train_loader_creator,
                       validation_data=validation_data_creator,
                       batch_size=cfg.TRAIN.BATCH_SIZE,
                       epochs=cfg.SOLVER.MAX_EPOCH,
                       recollate_fn=reduceIterElement)
```

### 4. How to test?
- [ ] N/A
- [x] Unit test
- [ ] Application test
- [ ] Document test
- [ ] ...

### CODE REVIEW

File: python/orca/src/bigdl/orca/learn/pytorch/pytorch_pyspark_estimator.py [hunk 1]
```diff
@@ -186,7 +186,8 @@ def fit(self,
             feature_cols=None,
             label_cols=None,
             validation_data=None,
-            callbacks=[]):
+            callbacks=[],
+            recollate_fn=empty_recollate_fn):
```
Comment (jason-dai) [inline python/orca/src/bigdl/orca/learn/pytorch/pytorch_pyspark_estimator.py:190]: shall simply call it `collate_fn`? @hkvision

Comment (hkvision) [reply python/orca/src/bigdl/orca/learn/pytorch/pytorch_pyspark_estimator.py:190]: Yes, I think `collate_fn` is better.

**Comment by hkvision** [comment]
Is the current behavior like this? If the data creator returns a DataLoader, then collate_fn in fit won't take effect since user is recommended to add it in its own dataloader.

**Comment by leonardozcm** [comment]
> Is the current behavior like this? If the data creator returns a DataLoader, then collate_fn in fit won't take effect since user is recommended to add it in its own dataloader.

I suppose not, `recollate_fn` is meant to align differences between dataloader and model from highly integrated codebase(slowfast for example)

That is to say, if the model or dataloader defined is on the user side, it is recommended to add it in their own code, but if it is not that convenient  to modify definition code, we  provide the user with one more chance to collate, that is why call it `re-collate`.

**Comment by leonardozcm** [comment]
Current logic is like this:
1. For user's dataloader, we wrap its `collate_fn` with `recollate_fn`:
```
data.collate_fn = dataloader_collate_wrapper(data.collate_fn, recollate_fn)
```
2. For user's dataframe/Xshard/Raydataset, `recollate_fn` will be applied on the dataloader which is constructed automatically。

**Pull request closed without merging**